src/main/java/com/google/idea/perf/vfstracer/VirtualFileTree.kt

```diff
@@ -0,0 +1,253 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.idea.perf.vfstracer
+
+import java.io.File
+import java.util.*
+
+/** A tree containing VirtualFile statistics. */
+interface VirtualFileTree {
+    companion object {
+        val EMPTY: VirtualFileTree = MutableVirtualFileTree.createRoot()
+    }
+
+    val name: String
+    val stubIndexAccesses: Int
+    val psiElementWraps: Int
+    val children: Map<String, VirtualFileTree>
+
+    val isDirectory: Boolean get() = children.isNotEmpty()
+    val isFile: Boolean get() = children.isEmpty()
+
+    fun statEquals(other: VirtualFileTree): Boolean =
+        stubIndexAccesses == other.stubIndexAccesses && psiElementWraps == other.psiElementWraps
+}
+
+class MutableVirtualFileTree(
+    override val name: String
+): VirtualFileTree {
+    companion object {
+        fun createRoot() = MutableVirtualFileTree("[root]")
+    }
+
+    override var stubIndexAccesses: Int = 0
+    override var psiElementWraps: Int = 0
+    override val children: MutableMap<String, MutableVirtualFileTree> = TreeMap()
+
+    /** Zeroes out all stat values. */
+    fun clear() {
+        stubIndexAccesses = 0
+        psiElementWraps = 0
+        for (child in children.values) {
+            child.clear()
+        }
+    }
+
+    /** Accumulates stat values from every node in [tree]. */
+    fun accumulate(tree: VirtualFileTree) {
+        for ((childName, child) in tree.children) {
+            val thisChild = children.getOrPut(childName) { MutableVirtualFileTree(childName) }
+            thisChild.accumulate(child)
+        }
+        stubIndexAccesses += tree.stubIndexAccesses
+        psiElementWraps += tree.psiElementWraps
+    }
+
+    /** Accumulates stat values from a given file path. */
+    fun accumulate(
+        path: String,
+        stubIndexAccesses: Int = 0,
+        psiElementWraps: Int = 0
+    ) {
+        val parts = getParts(path)
+        var tree = this
+        tree.stubIndexAccesses += stubIndexAccesses
+        tree.psiElementWraps += psiElementWraps
+
+        for (part in parts) {
+            val child = tree.children.getOrPut(part) { MutableVirtualFileTree(part) }
+            child.stubIndexAccesses += stubIndexAccesses
+            child.psiElementWraps += psiElementWraps
+            tree = child
+        }
+    }
+
+    private fun getParts(path: String): List<String> {
+        var file = File(path)
+        val parts = mutableListOf<String>()
+        do {
+            val name = file.name
+            parts.add(name)
+            file = file.parentFile
+        }
+        while (file.parentFile != null)
+
+        return parts.reversed()
```

Comment:
Could this function simplified to the following (and then inlined)?
```
return path.split(File.separatorChar)
```
(Note, even `file.parentFile` just uses `File.separatorChar` to find out where to split the path)

Reply:
👍 Yup, looking at the implementation of `file.getParentFile`, I can simplify this into a split.